tests/bit_view_tests.cpp

```diff
@@ -0,0 +1,417 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2022-2022  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "bit_view.h"
+
+#include <gtest/gtest.h>
+#include <memory>
+
+namespace {
+
+union Register {
+	uint8_t data = 0;
+	bit_view<0, 2> first_2;  // is bits 0 and 1
+	bit_view<2, 3> middle_3; // is bits 2, 3, and 4
+	bit_view<5, 3> last_3;   // is bits 5, 6, and 7
+};
```

Comment:
Nice 👍🏻

Reply:
Thanks!

It's about as far I think we can go and keep it in a very generic and clean C++ form.

It's now more convenient and safe, in that the `bit_view` will always use the correct `uint*_t` to fully cover the view. 

Compile-time checks guarantee that illegal views (negative index, zero or negative width, and views that overhang the start or end) aren't compiable - so we've got those hard safety bounds too. I've added unit tests to check these.

For the PC Speaker PR, I'll put this to use managing one of the 8-bit keyboard/PC-speaker registers (PPI controller port B register).

@shermp , shout if you see any gotchas; will hold off a bit before merging. Thanks guys :rocket:

Reply:
Great stuff, I agree, this strikes a good balance between usability/safety and simplicity of implementation.

It will be handy for my upcoming AdLib Gold stuff as well. I got it already working in a rather hacky way and it sounds glorious 😁

Reply:
> It will be handy for my upcoming AdLib Gold stuff as well. I got it already working in a rather hacky way and it sounds glorious :grin:

:exploding_head: :heart_eyes: .